### ISSUE TITLE
"Insert new child tabs as first tab" should ignore serial opening

### ISSUE THREAD

**Issue by pintassilgo** [piroor/treestyletab#1241]
I know that probably won't fix because of https://github.com/piroor/treestyletab/blob/master/README.md, but I'll try:

## Steps to reproduce

I use the following setting of Tree Style Tabs:
_Insertion position of new child tabs -> Insert to the top of the tree, as the first child_
Because I think that makes easier to switch to new tabs. But in case of open several tabs at once, I think this shouldn't be considered strictly.

Example:

Let's suppose I open a news portal, then open the Politics section in a new tab, as child. I'll keep both opened to update hourly.

After one update on the parent tab, there are 5 news that interests me, so I click all of them with middle button to open in background tabs.

## Expected result

. News portal
-- new tab 1
-- new tab 2
-- new tab 3
-- new tab 4
-- new tab 5
-- Politics section

This makes sense considering that "new tab 1" will be the first one to load, then I press Ctrl+Tab only once to go read the news without need do wait other tabs load nor having to go through the "Politics section" tab (this would happen if I changed "Insertion position of new child tabs" setting).

## Actual result

. News portal
-- new tab 5
-- new tab 4
-- new tab 3
-- new tab 2
-- new tab 1
-- Politics section

Thank you and forgive me if you find it invalid.

**Comment by pintassilgo** [comment]
The default behavior of Firefox (without TST) is as I expected.

Without TST, with _browser.tabs.insertRelatedAfterCurrent = true_, related tabs are inserted next to the "parent" tab, but in case of multiple new related tabs in a row, these are inserted as 1-2-3-4-5, which is right.

But with TST and _extensions.treestyletab.insertNewChildAt = 0_, they are inserted as 5-4-3-2-1.

I think I got a workaround, but certainly not in the proper way because I have no skill:

browser.js, line 3324
`newIndex = parent._tPos + 1;`
replaced by
```
var momt = parent._tPos;
for (let a = 0; a < gBrowser.mTabs.length; a++) {
	if (gBrowser.mTabs[a].owner === undefined) {
		momt = gBrowser.mTabs[a]._tPos;
		break;
	}
}
newIndex = momt._tPos + 1;
```

This works because the most recent child tab opened in background doesn't have the _owner_ attribute set until I select/close another tab.

**Comment by piroor** [comment]
Because the behavior you told is same to Firefox's default one, I've introduced a new choice for the position of new child tab: "no control".

 * 54cf06daa390f524b3c84fd015c82bb67e6727b4
 * 6ae91a7d18dd94a7639fa5ef1ac29a92659ad45b

I think this option helps you.

**Comment by pintassilgo** [comment]
Thank you, but this broke opening child tabs (inside parent tree). Only the last tab of tabbar is able to open child tabs, the others don't.

https://i.imgur.com/cELFp1o.gif

Firefox 52, no other extension enabled.

**Comment by piroor** [comment]
Thank you for the feedback. Does d83d2e9acfdef31edf8c07db1d48fa48a9dfd054 work?

**Comment by pintassilgo** [comment]
On a fresh profile, yes.

On my main, with TabGroups Manager Revived, I can only open child tab in one group, usually the last. If I create a new group, it takes de condition of being the only one that works as it should.

Since "insert as first child" has no conflict with TabGroups Manager Revived, this means that TST has the final word, so if TST could have the default Firefox behavior hardcoded (and not "behave as Firefox is set up"), I think this will solve the problem.

But I understand if you consider this to be very particular and fault of another extension. Thank you anyway.

**Comment by piroor** [comment]
Hmm, I've tried TabGroups Manager Revived 2016.02.03.008 with TST master/HEAD and the result is:

 * In the initial group, new tabs are become children of the current tab as expected.
 * In the second group, tabs are not organized. (unexpected)
 * After I open new third group, tabs are organized. 4th and 5th groups also.

**Comment by pintassilgo** [comment]
This is because the third, 4th and 5th were always the last group for you. If you switch back to one of the previous groups without closing the last one and try again, you'll see that things will not be as expected.

First case:
- let tab of first group without child
- realizing that your last group are OK, switch back to first group and try to open some links at once
- new tabs will not be organized, but the order will be as expected (1, 2, 3...)

Second case:
- let tab of first group with at least ore child
- realizing that your last group are OK, switch back to first group and try to open some links at once
- new tabs will be children, but the order will be wrong (...3, 2, 1)

About your initial group being as expected and the second not, maybe internally TGMR considered that the initial group was in fact the last, even if the order of TGMR's bar didn't show that.

**Comment by piroor** [comment]
OK, I got it.

```
[treestyletab:browser] detachTab: detach 2 from 0  utils_logWithStackTrace@resource://treestyletab-modules/utils.js:303:16
  logWithStackTrace@resource://treestyletab-modules/browser.js:95:2
  TSTBrowser_detachTab@resource://treestyletab-modules/browser.js:5852:3
  TSTBrowser_partTab@resource://treestyletab-modules/browser.js:5899:10
  TabGroupsManager.tabMoveByTGM.moveTabToWithoutTST@chrome://tabgroupsmanager/content/TabGroupsManager.js:361:7
  TabGroupsManager.GroupClass.prototype.moveTabToLast@chrome://tabgroupsmanager/content/TabGroupsManager.js:3265:5
  TabGroupsManager.GroupClass.prototype.addTabToTabArray@chrome://tabgroupsmanager/content/TabGroupsManager.js:3083:9
  TabGroupsManager.GroupClass.prototype.addTab@chrome://tabgroupsmanager/content/TabGroupsManager.js:3062:7
  TabGroupsManager.EventListener.prototype.onTabOpen@chrome://tabgroupsmanager/content/TabGroupsManager.js:1663:11
  TabGroupsManager.EventListener.prototype.handleEvent@chrome://tabgroupsmanager/content/TabGroupsManager.js:1626:19
  addTab@chrome://browser/content/tabbrowser.xml:2250:13
  TabGroupsManager.OverrideMethod.prototype.override_gBrowser_addTab@chrome://tabgroupsmanager/content/TabGroupsManager.js:5679:10
  loadOneTab@chrome://browser/content/tabbrowser.xml:1564:23
  TabGroupsManager.OverrideMethod.prototype.override_gBrowser_loadOneTab@chrome://tabgroupsmanager/content/TabGroupsManager.js:5687:10
  openLinkIn@chrome://browser/content/utilityOverlay.js:393:26
  TSTWH_preInit/window.openLinkIn@chrome://treestyletab/content/windowHelper.js:99:80
  handleLinkClick@chrome://browser/content/browser.js:5623:3
  TSTWH_preInit/window.handleLinkClick@chrome://treestyletab/content/windowHelper.js:113:88
  contentAreaClick@chrome://browser/content/browser.js:5545:3
```

TGMR calls TST's API `gBrowser.treeStyleTab.partTab(tab);` to detach a newly opened tab from its parent. It breaks the tree...

**Comment by piroor** [comment]
The main topic of this issue is done, so I separate the compatibility issue with TGMR as a new independent issue.
